collections/binary_search_tree.ts

```diff
@@ -235,33 +235,35 @@ export class BinarySearchTree<T> implements Iterable<T> {
     return null;
   }
 
+  /** Removes the given node, and returns the node that was physically removed from the tree. */
   protected removeNode(
-    value: T,
+    node: BinarySearchNode<T>,
   ): BinarySearchNode<T> | null {
-    let removeNode: BinarySearchNode<T> | null = this.findNode(value);
-    if (removeNode) {
-      const successorNode: BinarySearchNode<T> | null =
-        !removeNode.left || !removeNode.right
-          ? removeNode
-          : removeNode.findSuccessorNode()!;
-      const replacementNode: BinarySearchNode<T> | null = successorNode.left ??
-        successorNode.right;
-      if (replacementNode) replacementNode.parent = successorNode.parent;
+    /**
+     * The node to physically remove from the tree.
+     * Guaranteed to not have both left and right children.
+     */
+    const flaggedNode: BinarySearchNode<T> | null = !node.left || !node.right
```

Comment:
Changed the naming here because the node that we are removing is not always the "successor" node, and I feel the naming here has caused confusion.

The node flagged for removal should only be the successor node when the node we're trying to remove has two children. Otherwise, we should be removing the provided node.